### ISSUE TITLE
MismatchedInputException: @JsonCreator ignored with ParameterNamesModule and Getter

### ISSUE THREAD

**Issue by xathien** [FasterXML/jackson-databind#2112]
Similar to #1318 and #1924 but (perhaps significantly?) different reproduction method.

Tested on `2.9.1` and `2.9.6` (currently latest). Works without ParameterNamesModule registered (or if compiled without `-parameters`, of course!)

```java
import java.util.UUID;

public class JsonCreatorTest
{
  @Test
  public void deleteMe() throws Exception
  {
    new ObjectMapper()
        .registerModule(new ParameterNamesModule())
        .readValue("\"6994179a-11ff-4fdd-969c-bc977b50332d\"", UuidWrapper.class);
  }

  public static class UuidWrapper
  {
    private final UUID val;

    public UUID getVal()
    {
      return val;
    }

    private UuidWrapper(UUID val)
    {
      this.val = val;
    }

    @JsonCreator
    public static UuidWrapper valueOf(String val)
    {
      return new UuidWrapper(UUID.fromString(val));
    }
  }
}
```

Causes:
```
com.fasterxml.jackson.databind.exc.MismatchedInputException: 
Cannot construct instance of `foo.JsonCreatorTest$UuidWrapper` (although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('6994179a-11ff-4fdd-969c-bc977b50332d')
 at [Source: (String)""6994179a-11ff-4fdd-969c-bc977b50332d""; line: 1, column: 1]
	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:63)
	at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1342)
	at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1031)
	at com.fasterxml.jackson.databind.deser.ValueInstantiator._createFromStringFallbacks(ValueInstantiator.java:371)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromString(StdValueInstantiator.java:323)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromString(BeanDeserializerBase.java:1373)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:171)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:161)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4013)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3004)
```

**Comment by cowtowncoder** [comment]
Thank you for reporting this, and especially for providing reproduction. I will try to modify it so it can be tested without parameter names module, with mock parameter name provider.

**Comment by xathien** [comment]
Thanks for getting back in record time!

In putting together the repro case, I've discovered a workaround, too. Renaming the param for the `valueOf` method will fix it, e.g.
```java
    @JsonCreator
    public static UuidWrapper valueOf(String valString)
    {
      return new UuidWrapper(UUID.fromString(valString));
    }
```

**Comment by cowtowncoder** [comment]
@xathien I must admit response time was lucky coincidence :-D

Another workaround is to specify mode explicitly with:

    @JsonCreator(mode = JsonCreator.Mode.DELEGATING)

but this all actually points to a significant problem -- since this goes back to heuristic used to determine mode (delegating vs properties-based), I am not sure I can actually change it at least for 2.9.
Problem being that it is possible that other users might be relying on this as a feature.

Specifically what happens is that due to existence of that `val` field AND parameter name `val` (taken as implicit name), Jackson guesses that intent is for this to match content like:

```json
{
   "val" : "6994179a-11ff-4fdd-969c-bc977b50332d"
}
```

and not

```json
"6994179a-11ff-4fdd-969c-bc977b50332d"
```

and this is where exception comes from.

**Comment by cowtowncoder** [comment]
Come to think of this now, there is yet another work-around: you can also make parameter name module default to mode "delegating" as well:

```java
               mapper.registerModule(new ParameterNamesModule(JsonCreator.Mode.DELEGATING));
```

which will then decide on potentially ambiguous case by declaring it should be "delegating", without using heuristics.

So. One of these options is your best bet at this point. I don't think I can change the behavior for 2.9, but even for 2.10 it is unclear whether handling can be improved -- I could consider defaulting heuristics to only apply to constructors, not factory methods, but that is likely to be no less confusing as well as to probably break some existing usage.

**Comment by xathien** [comment]
I actually tried `mode = DELEGATING` from comments on other bugs you made, but it didn't fix the problem in my codebase (not sure whether it'd workaround for the stripped-down repro steps I gave you). Also tried disabling a couple of mapper features that you had suggested, and those didn't seem to affect much of anything. But I'll stick with renaming the param for the moment, and hope for a bug fix in future Jackson versions that I'll eagerly upgrade to! 😄
